### ISSUE TITLE
Token Level Embeddings from Universal Sentence Encoder

### ISSUE THREAD

**Issue by amankedia** [tensorflow/hub#403]
Hello,
I was trying to extract token level embeddings from Universal Sentence Encoder(USE). The paper on the same here says https://arxiv.org/pdf/1803.11175.pdf that the token level embeddings are normalized using square root of the length of the sentence.

However, when I tried to print the model variables available on TensorFlow Hub, they give me a form as pasted in the attached code block at the end. I am unable to figure out which layer would exactly correspond to token level embeddings.

Anything on this would be highly appreciated.

`[<tf.Variable 'module/Embeddings_en/sharded_0:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_1:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_10:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_11:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_12:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_13:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_14:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_15:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_16:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_2:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_3:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_4:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_5:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_6:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_7:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_8:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Embeddings_en/sharded_9:0' shape=(47060, 320) dtype=float32_ref>,
 <tf.Variable 'module/Encoder_en/DNN/ResidualHidden_0/weights:0' shape=(320, 320) dtype=float32_ref>,
 <tf.Variable 'module/Encoder_en/DNN/ResidualHidden_1/weights:0' shape=(320, 320) dtype=float32_ref>,
 <tf.Variable 'module/Encoder_en/DNN/ResidualHidden_2/weights:0' shape=(320, 512) dtype=float32_ref>,
 <tf.Variable 'module/Encoder_en/DNN/ResidualHidden_3/projection:0' shape=(320, 512) dtype=float32_ref>,
 <tf.Variable 'module/Encoder_en/DNN/ResidualHidden_3/weights:0' shape=(512, 512) dtype=float32_ref>,
 <tf.Variable 'module/SHARED_RANK_ANSWER/response_encoder_0/tanh_layer_0/bias:0' shape=(320,) dtype=float32_ref>,
 <tf.Variable 'module/SHARED_RANK_ANSWER/response_encoder_0/tanh_layer_0/weights:0' shape=(512, 320) dtype=float32_ref>,
 <tf.Variable 'module/SHARED_RANK_ANSWER/response_encoder_0/tanh_layer_1/bias:0' shape=(320,) dtype=float32_ref>,
 <tf.Variable 'module/SHARED_RANK_ANSWER/response_encoder_0/tanh_layer_1/weights:0' shape=(320, 320) dtype=float32_ref>,
 <tf.Variable 'module/SHARED_RANK_ANSWER/response_encoder_0/tanh_layer_2/bias:0' shape=(512,) dtype=float32_ref>,
 <tf.Variable 'module/SHARED_RANK_ANSWER/response_encoder_0/tanh_layer_2/weights:0' shape=(320, 512) dtype=float32_ref>,
 <tf.Variable 'module/SNLI/Classifier/LinearLayer/bias:0' shape=(3,) dtype=float32_ref>,
 <tf.Variable 'module/SNLI/Classifier/LinearLayer/weights:0' shape=(512, 3) dtype=float32_ref>,
 <tf.Variable 'module/SNLI/Classifier/tanh_layer_0/bias:0' shape=(512,) dtype=float32_ref>,
 <tf.Variable 'module/SNLI/Classifier/tanh_layer_0/weights:0' shape=(2048, 512) dtype=float32_ref>,
 <tf.Variable 'module/global_step:0' shape=() dtype=int32_ref>]`

**Comment by iamxpy** [comment]
@amankedia @rmothukuru Hello guys, have you found the method to extract token level / word level representation from USE? I want to get the word level vectors from pre-trained models from tensorflow-hub.Though [elmo](https://tfhub.dev/google/elmo/3) may help, I am afraid that LSTMs in elmo may be too slow for me, thus I am finding the way to get word embeddings from USE( or other models ) which contains relatively faster self-attention operation. If you know the solution, please tell me! Thanks!

**Comment by gowthamkpr** [comment]
Hi

The variables 'module/Embeddings_en/sharded_* contains the token level embeddings. But in the graph, there are two steps to get the token level embeddings:
1. at the first step, a in graph tokenizer tokenize the input text to unigram and bigram tokens. each of the token is mapped to a id using a mapping function
2. for each token id, a embedding_look function fetches the token (id) embedding from the embedding variable tensors.

After this step, (in DNN module) all token level embeddings are sum and normalized as described in paper.

-As per Yinfieys explanation.
